src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Remote/ServiceDescriptors.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.ServiceHub.Framework;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor
+{
+    /// <summary>
+    /// Service descriptors of brokered Roslyn ServiceHub services.
+    /// </summary>
+    internal static class ServiceDescriptors
+    {
+        /// <summary>
+        /// Brokered services must be defined in Microsoft.VisualStudio service namespace in order to be considered first party.
+        /// </summary>
+        internal const string ServiceNamePrefix = "Microsoft.VisualStudio.Razor.";
+
+        private const string InterfaceNamePrefix = "IRemote";
+        private const string InterfaceNameSuffix = "Service";
+
+        internal static readonly ImmutableDictionary<Type, ServiceDescriptor> Descriptors = ImmutableDictionary.CreateRange(new[]
+        {
+            CreateDescriptor(typeof(IRemoteLanguageService), callbackInterface: null),
+        });
+
+        private static string GetServiceName(Type serviceInterface)
+        {
+            Debug.Assert(serviceInterface.IsInterface);
+            var interfaceName = serviceInterface.Name;
+            Debug.Assert(interfaceName.StartsWith(InterfaceNamePrefix, StringComparison.Ordinal));
+            Debug.Assert(interfaceName.EndsWith(InterfaceNameSuffix, StringComparison.Ordinal));
+
+            return ServiceNamePrefix + interfaceName.Substring(InterfaceNamePrefix.Length, interfaceName.Length - InterfaceNamePrefix.Length - InterfaceNameSuffix.Length);
+        }
+
+        private static KeyValuePair<Type, ServiceDescriptor> CreateDescriptor(Type serviceInterface, Type? callbackInterface = null)
+        {
+            Debug.Assert(callbackInterface == null || callbackInterface.IsInterface);
+
+            var serviceName = GetServiceName(serviceInterface);
+            var descriptor = ServiceDescriptor.CreateRemoteServiceDescriptor(serviceName, callbackInterface);
```

Comment:
What's the purpose of a callback interface?

Reply:
You're not using it currently, but you can define an interface that can be used in the service to call back to the client.